python_modules/dagster/dagster/_core/storage/runs/in_memory.py

```diff
@@ -1,410 +1,54 @@
-from collections import OrderedDict, defaultdict
-from typing import Callable, Dict, Iterable, List, Optional, Set, Tuple, Union, cast
+from contextlib import contextmanager
 
-import dagster._check as check
-from dagster._core.errors import (
-    DagsterRunAlreadyExists,
-    DagsterRunNotFoundError,
-    DagsterSnapshotDoesNotExist,
-)
-from dagster._core.events import EVENT_TYPE_TO_PIPELINE_RUN_STATUS, DagsterEvent, DagsterEventType
-from dagster._core.execution.backfill import BulkActionStatus, PartitionBackfill
-from dagster._core.snap import (
-    ExecutionPlanSnapshot,
-    PipelineSnapshot,
-    create_execution_plan_snapshot_id,
-    create_pipeline_snapshot_id,
-)
-from dagster._core.storage.tags import PARTITION_NAME_TAG
-from dagster._daemon.types import DaemonHeartbeat
-from dagster._utils import EPOCH, frozendict, merge_dicts
+import sqlalchemy as db
+from sqlalchemy.pool import NullPool
 
-from ..pipeline_run import (
-    JobBucket,
-    PipelineRun,
-    RunPartitionData,
-    RunRecord,
-    RunsFilter,
-    TagBucket,
-)
-from .base import RunStorage
+from dagster._core.storage.sql import create_engine, get_alembic_config, stamp_alembic_rev
+from dagster._core.storage.sqlite import create_in_memory_conn_string
 
+from .schema import InstanceInfo, RunStorageSqlMetadata
+from .sql_run_storage import SqlRunStorage
 
-def build_run_filter(filters: Optional[RunsFilter]) -> Callable[[PipelineRun], bool]:
-    def _filter(run: PipelineRun) -> bool:
-        if not filters:
-            return True
 
-        if filters.run_ids and run.run_id not in filters.run_ids:
-            return False
+class InMemoryRunStorage(SqlRunStorage):
+    """
+    In memory only run storage. Used by ephemeral DagsterInstance or for testing purposes.
 
-        if filters.statuses and run.status not in filters.statuses:
-            return False
+    WARNING: Dagit and other core functionality will not work if this is used on a real DagsterInstance
+    """
 
-        if filters.job_name and filters.job_name != run.pipeline_name:
-            return False
-
-        if filters.mode and filters.mode != run.mode:
-            return False
-
-        if filters.tags and not all(
-            (run.tags.get(key) == value if isinstance(value, str) else run.tags.get(key) in value)
-            for key, value in filters.tags.items()
-        ):
-            return False
-
-        if filters.snapshot_id and filters.snapshot_id != run.pipeline_snapshot_id:
-            return False
-
-        return True
-
-    return _filter
+    def __init__(self, preload=None):
+        self._conn = self._create_connection()
 
+        self.migrate()
+        self.optimize()
```

Comment:
? isnt this a fresh DB everytime?

Reply:
We have some unit tests in place that check to see if the migrations and some of the secondary index values are set